packages/flame/lib/src/collisions/quadtree/has_quadtree_collision_detection.dart

```diff
@@ -0,0 +1,100 @@
+import 'package:flame/collisions.dart';
+import 'package:flame/components.dart';
+import 'package:flame/extensions.dart';
+import 'package:flame/game.dart';
+
+/// Mixin should be applied to FlameGame to bring QuadTree collision support.
+/// [initializeCollisionDetection] should be called at [onLoad] function.
+///
+mixin HasQuadTreeCollisionDetection on FlameGame
+    implements HasCollisionDetection<QuadTreeBroadphase<ShapeHitbox>> {
+  late QuadTreeCollisionDetection _collisionDetection;
+
+  @override
+  QuadTreeCollisionDetection get collisionDetection => _collisionDetection;
+
+  @override
+  set collisionDetection(
+    CollisionDetection<ShapeHitbox, QuadTreeBroadphase<ShapeHitbox>> cd,
+  ) {
+    if (cd is! QuadTreeCollisionDetection) {
+      throw 'Must be QuadTreeCollisionDetection!';
+    }
+    _collisionDetection = cd;
+  }
+
+  /// Initialise the QuadTree.
+  ///
+  /// - [mapDimensions] describes the collision area coordinates and size.
+  ///   Should match to game map position and size.
+  /// - [maxObjects] (optional) - maximum objects count in one quadrant.
+  /// - [maxLevels] (optional) - maximum nested quadrants.
+  /// - [minimumDistance] (optional) - specify minimum distance between objects
+  ///   to consider them as possibly colliding. You can also implement the
+  ///   [minimumDistanceCheck] if you need some custom behavior.
+  ///
+  /// The [onComponentTypeCheck] checks if objects of different types should
+  /// collide.
+  /// The result of the calculation is cached so you should not check any
+  /// dynamical parameters here, the function is intended to be used as pure
+  /// type checker.
+  /// It usually should not be overridden, see
+  /// [CollisionCallbacks.onComponentTypeCheck] instead
+  void initializeCollisionDetection({
+    required Rect mapDimensions,
+    double? minimumDistance,
+    int maxObjects = 25,
+    int maxLevels = 10,
+  }) {
+    _collisionDetection = QuadTreeCollisionDetection(
+      mapDimensions: mapDimensions,
+      maxDepth: maxLevels,
+      maxObjects: maxObjects,
+      onComponentTypeCheck: onComponentTypeCheck,
+      minimumDistanceCheck: minimumDistanceCheck,
+    );
+    this.minimumDistance = minimumDistance;
+  }
+
+  double? minimumDistance;
+
+  bool minimumDistanceCheck(Vector2 activeItemCenter, Vector2 potentialCenter) {
+    return minimumDistance == null ||
+        !((activeItemCenter.x - potentialCenter.x).abs() > minimumDistance! ||
+            (activeItemCenter.y - potentialCenter.y).abs() > minimumDistance!);
+  }
+
+  bool onComponentTypeCheck(PositionComponent one, PositionComponent another) {
```

Comment:
Aren't these always `ShapeHitbox`es? Then both this and the collision callback can be changed to that, and then the collision callback should use `hitboxParent` instead of `parent`.

Reply:
I can't replace types at function declaration because if I pass "parent" here, it would not to be a ShapeHitbox.

Reply:
Ah, of course, my bad.